File: src/test/java/de/fraunhofer/aisec/cpg/frontends/cpp/CXXSymbolConfigurationTest.java
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */

package de.fraunhofer.aisec.cpg.frontends.cpp;

import static de.fraunhofer.aisec.cpg.TestUtils.analyzeAndGetFirstTU;
import static de.fraunhofer.aisec.cpg.TestUtils.analyzeWithBuilder;
import static org.junit.jupiter.api.Assertions.*;

import de.fraunhofer.aisec.cpg.BaseTest;
import de.fraunhofer.aisec.cpg.TranslationConfiguration;
import de.fraunhofer.aisec.cpg.graph.*;
import java.io.File;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.jupiter.api.Test;

class CXXSymbolConfigurationTest extends BaseTest {
  @Test
  void testWithoutSymbols() throws Exception {
    File file = new File("src/test/resources/symbols.cpp");
    // parse without symbols
    TranslationUnitDeclaration tu =
        analyzeAndGetFirstTU(List.of(file), file.getParentFile().toPath(), true);

    Set<FunctionDeclaration> main = tu.getDeclarationsByName("main", FunctionDeclaration.class);
    assertFalse(main.isEmpty());
    FunctionDeclaration funcDecl = main.iterator().next();

    BinaryOperator binaryOperator = funcDecl.getBodyStatementAs(0, BinaryOperator.class);
    assertNotNull(binaryOperator);

    // without additional symbols, the first line will look like a reference (to something we do not
    // know)
    DeclaredReferenceExpression dre = binaryOperator.getRhsAs(DeclaredReferenceExpression.class);
    assertNotNull(dre);
    assertEquals("HELLO_WORLD", dre.getName());

    binaryOperator = funcDecl.getBodyStatementAs(1, BinaryOperator.class);
    assertNotNull(binaryOperator);

    // without additional symbols, the second line will look like a function call (to something we
    // do not know)
    CallExpression call = binaryOperator.getRhsAs(CallExpression.class);
    assertNotNull(call);
    assertEquals("INCREASE", call.getName());
  }

  @Test
  void testWithSymbols() throws Exception {
    File file = new File("src/test/resources/symbols.cpp");

    TranslationUnitDeclaration tu =
        analyzeWithBuilder(
                TranslationConfiguration.builder()
                    .sourceLocations(List.of(file))
                    .topLevel(file.getParentFile())
                    .symbols(
                        Map.of(
                            "HELLO_WORLD", "\"Hello World\"",
                            "INCREASE(X)", "X+1"))
                    .defaultPasses())
            .iterator()
            .next();

    Set<FunctionDeclaration> main = tu.getDeclarationsByName("main", FunctionDeclaration.class);
    assertFalse(main.isEmpty());
    FunctionDeclaration funcDecl = main.iterator().next();

    BinaryOperator binaryOperator = funcDecl.getBodyStatementAs(0, BinaryOperator.class);
    assertNotNull(binaryOperator);

    // should be a literal now
    Literal<?> literal = binaryOperator.getRhsAs(Literal.class);
    assertEquals("Hello World", literal.getValue());

    binaryOperator = funcDecl.getBodyStatementAs(1, BinaryOperator.class);
    assertNotNull(binaryOperator);

    // should be expanded to another binary operation 1+1
    BinaryOperator add = binaryOperator.getRhsAs(BinaryOperator.class);
    assertNotNull(add);
    assertEquals("+", add.getOperatorCode());

    Literal<?> literal2 = add.getLhsAs(Literal.class);
    assertEquals(2, literal2.getValue());

    Literal<?> literal1 = add.getRhsAs(Literal.class);
    assertEquals(1, literal1.getValue());
  }
}

File: cpg-language-cxx/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CXXSymbolConfigurationTest.kt
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.frontends.cxx

import de.fraunhofer.aisec.cpg.*
import de.fraunhofer.aisec.cpg.frontends.TranslationException
import de.fraunhofer.aisec.cpg.graph.*
import de.fraunhofer.aisec.cpg.graph.statements.expressions.BinaryOperator
import de.fraunhofer.aisec.cpg.graph.statements.expressions.CallExpression
import de.fraunhofer.aisec.cpg.graph.statements.expressions.Literal
import de.fraunhofer.aisec.cpg.graph.statements.expressions.Reference
import de.fraunhofer.aisec.cpg.test.*
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

internal class CXXSymbolConfigurationTest : BaseTest() {
    @Test
    @Throws(TranslationException::class)
    fun testWithoutSymbols() {
        val ctx = TranslationContext(TranslationConfiguration.builder().build())

        // parse without symbols
        val tu =
            CXXLanguageFrontend(ctx, CPPLanguage(ctx)).parse(File("src/test/resources/symbols.cpp"))
        val main = tu.functions["main"]
        assertNotNull(main)

        val funcDecl = main
        var binaryOperator = funcDecl.bodyOrNull<BinaryOperator>(0)
        assertNotNull(binaryOperator)

        // without additional symbols, the first line will look like a reference (to something we do
        // not know)
        val dre = binaryOperator.rhs<Reference>()
        assertNotNull(dre)
        assertLocalName("HELLO_WORLD", dre)

        binaryOperator = funcDecl.bodyOrNull<BinaryOperator>(1)
        assertNotNull(binaryOperator)

        // without additional symbols, the second line will look like a function call (to something
        // we do not know)
        val call = binaryOperator.rhs<CallExpression>()
        assertNotNull(call)
        assertLocalName("INCREASE", call)
    }

    @Test
    @Throws(TranslationException::class)
    fun testWithSymbols() {
        val config =
            TranslationConfiguration.builder()
                .symbols(mapOf(Pair("HELLO_WORLD", "\"Hello World\""), Pair("INCREASE(X)", "X+1")))
                .defaultPasses()
                .build()

        // let's try with symbol definitions
        val ctx = TranslationContext(config)
        val tu =
            CXXLanguageFrontend(ctx, CPPLanguage(ctx)).parse(File("src/test/resources/symbols.cpp"))
        val main = tu.functions["main"]
        assertNotNull(main)

        val funcDecl = main
        var binaryOperator = funcDecl.bodyOrNull<BinaryOperator>(0)
        assertNotNull(binaryOperator)

        // should be a literal now
        val literal = binaryOperator.rhs<Literal<String>>()
        assertNotNull(literal)
        assertEquals("Hello World", literal.value)

        binaryOperator = funcDecl.bodyOrNull<BinaryOperator>(1)
        assertNotNull(binaryOperator)

        // should be expanded to another binary operation 1+1
        val add = binaryOperator.rhs<BinaryOperator>()
        assertNotNull(add)
        assertEquals("+", add.operatorCode)

        val literal2 = add.lhs<Literal<Int>>()
        assertNotNull(literal2)
        assertEquals(2, literal2.value)

        val literal1 = add.rhs<Literal<Int>>()
        assertNotNull(literal1)
        assertEquals(1, literal1.value)
    }
}
